### PR TITLE
AVRO-2703: Use KMP Algorithm For Sync Marker Search

### PR DESCRIPTION
**Pull request by belugabehr**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [AVRO-2703](https://issues.apache.org/jira/browse/AVRO/AVRO-2703) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-2703
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [X] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Comment by rstata** [comment]
I wouldn't be in favor of merging this change.

Sync's don't happen that often, when they do it's not performance critical -- but correctness is important.  Also, the likelihood of getting more than 2-3 bytes into the sync marker and NOT syncing is low (ie, the naive algo is not likely to see worst-case performance).

Putting all of this together, the existing, naive algo is the better choice, i.e., an algo whose correctness is easier to check by inspection even if it's (theoretically) slower.

**Comment by belugabehr** [comment]
As I understand it, sync is most often used when starting splits for
MR/Spark.  Using a more efficient algorithm would allow flexibility for
longer gaps between syncs and potentially faster time to start processing
data.

KMP is a very well understood and often used algorithm for string search.
As you can see, it is roughly 20 lines of code.  If you would like
additional unit tests, I can add, but the existing suite passes as-is.

I can also provide some performance data as well.

**Comment by belugabehr** [comment]
Also, the 'i' counter here could overflow on a very large file.

**Comment by belugabehr** [comment]
First batch of testing is complete:

```
# Run complete. Total time: 00:10:08

Benchmark                   Mode  Cnt     Score     Error  Units
BenchmarkSearch.testKmp    thrpt   50  5223.917 ± 579.125  ops/s
BenchmarkSearch.testNaive  thrpt   50  2775.205 ± 149.943  ops/s
```
I create a 64K buffer with random data, and filled it with random data, the 16b sequence was also randomly generated.  This simulates the worst-case scenario in that the sequence value is not found. This setup requires the algos to read all the way through the data.

File: lang/java/avro/src/main/java/org/apache/avro/file/DataFileReader.java [hunk 1]
```diff
@@ -143,36 +143,61 @@ public void seek(long position) throws IOException {
...
-    } catch (EOFException e) {
-      // fall through
+
+    if (this.partialMatchTable == null) {
+      this.partialMatchTable = computePartialMatchTable(getHeader().sync);
     }
-    // if no match or EOF set start to the end position
+
+    final byte[] sync = getHeader().sync;
+    final InputStream in = vin.inputStream();
```
Comment (Fokko) [inline lang/java/avro/src/main/java/org/apache/avro/file/DataFileReader.java]: Shouldn't we close this `InputStream`?

Comment (belugabehr) [reply lang/java/avro/src/main/java/org/apache/avro/file/DataFileReader.java]: @Fokko I don't think so.  I think the idea is to position the stream to the right place and then start reading the stream in subsequent calls.  The current implementation also behaves this way.

**Comment by belugabehr** [comment]
I just updated the partial match table implementation with something I think it a small bit cleaner.  I also cited where I took it from.

For a sanity check, I took the two implementations (the first PR and this most recent one) and had them generate billions of random sequences and then compare the two results together.  All samples produced the same partial match table.

**Comment by belugabehr** [comment]
The current default sync one sync marker per 64K.  So my last test was 'worst case' scenario where the sync method would have to look at the entire block.  However, in practice, throwing a dart, on average, the sync method would have to read 32K bytes.  Also, the sync method probably gets called once per task, just to start the read point at the start of a sync point.  So, I created a new JMH test to work on a 32K buffer and a new DataFileReader was created for each sync attempt (in that way, the KMP partial match table is computed every time the sync method is invoked and the fixed-cost is measured each attempt).  Here are the results:

```
BenchmarkSearch.testKmp    thrpt   50  13354.194 ±  61.090  ops/s
BenchmarkSearch.testNaive  thrpt   50   5707.727 ± 119.155  ops/s
```

**Comment by belugabehr** [comment]
@cutting Any thoughts on this one?

**Pull request merged**